app/gui/view/graph-editor/src/lib.rs

```diff
@@ -1744,16 +1746,110 @@ impl GraphEditorModel {
+    ///
+    /// See [`new_node_position_aligned_if_close_to_node`] for details on what "close to" means.
+    pub fn new_node_position_at_mouse_aligned_to_source_node(
+        &self,
+        edge_id: EdgeId,
+        mouse_position: Vector2,
+    ) -> Vector2 {
+        let source_node_id = self.edge_source_node_id(edge_id);
+        let source_node = source_node_id.and_then(|id| self.nodes.get_cloned_ref(&id));
+        self.new_node_position_aligned_if_close_to_node(mouse_position, source_node)
+    }
+
+    /// Calculate a position for a new node, starting from the mouse position and aligning to the
+    /// node closest to the mouse position if it is close enough.
+    ///
+    /// See [`new_node_position_aligned_if_close_to_node`] for details on what is close enough.
+    pub fn new_node_position_at_mouse_aligned_to_close_nodes(
+        &self,
+        mouse_position: Vector2,
+    ) -> Vector2 {
+        let nearest_node = self.node_nearest_to_point(mouse_position);
+        self.new_node_position_aligned_if_close_to_node(mouse_position, nearest_node)
+    }
+
+    /// Return a node nearest to the specified point.
+    ///
+    /// The distance between a point and a node is the distance between the point and the node's
+    /// bounding box.
+    fn node_nearest_to_point(&self, point: Vector2) -> Option<Node> {
+        let mut min_distance_squared = f32::MAX;
+        let mut nearest_node = None;
+        let nodes = self.nodes.all.raw.borrow();
+        for node in nodes.values() {
+            let node_bounding_box = node.frp.bounding_box.value();
+            let distance_squared = node_bounding_box.squared_distance_to_point(point);
+            if distance_squared < min_distance_squared {
+                min_distance_squared = distance_squared;
+                nearest_node = Some(node.clone_ref());
+            }
+        }
+        nearest_node
+    }
+
+    /// Calculates a position for a new node, aligning it to the specified reference node if the
+    /// proposed position is close enough to it.
+    ///
+    /// A point is close enough to a node if it is located in an alignment area around a node,
+    /// defined in the current theme ([`theme::graph_editor::alignment_area_around_node`]).
+    ///
+    /// In the picture below, the solid border represents the node, while the dashed border
+    /// represents the alignment area. The captions are used as the variables in the code.
+    /// ```text
+    /// ┌┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┐
+    /// ┆               ▲               ┆
+    /// ┆           top │               ┆
+    /// ┆               ▼               ┆
+    /// ┆ left  ┌───────────────┐ right ┆
+    /// ┆ ◀───▶ └───────────────┘ ◀───▶ ┆
+    /// ┆               ▲               ┆
+    /// ┆        bottom │               ┆
+    /// ┆               ▼               ┆
+    /// └┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┘
+    /// ```
+    pub fn new_node_position_aligned_if_close_to_node(
```

Comment:
As agreed separately in a private talk on Discord, I plan to postpone this to a later PR, after the changes reverted in https://github.com/enso-org/enso/pull/3328 are re-applied with fixes. That is in order to avoid wasting time on laborious merge conflicts resolution (the proposed refactoring would move functions that are created and/or removed by #3328 - resolving such merge conflicts is certainly doable, but I believe it is a much more effective use of our time to postpone the refactoring).

Reply:
I don't understand what @MichaelMauderer question has to do with merge conflicts. You can do what Michael asked about without bumping develop, and bump it after #3328 is re-applied with fixes. 

I second what Michael asked about. Growing graph-eidotr/src/lib.rs is bad, and extracting it to a separate file seems like a very good idea.

Reply:
I moved the functions to a separate file. After the refactoring, I stripped the common prefix `new_node_position_` from the function names, to avoid repeating the module name in function names. If that makes the resulting names incorrect, I'd be grateful for help and suggestions on how to make them good!